# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2025-02-05T09:41:05.747255494+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkjhrao65/model/ with prefix 14e9a61947ef4edd
[INFO 2025-02-05T09:41:05.758990647+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:41:05.759105872+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2025-02-05T09:41:09.070196864+00:00 kernel.cc:1214] Loading model from path /tmp/tmpamhwvohr/model/ with prefix a234f4078ffe449c
[INFO 2025-02-05T09:41:09.081325619+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2025-02-05T09:41:09.081399487+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmptoc00a2p as temporary training directory


[INFO 2025-02-05T09:43:11.371013739+00:00 kernel.cc:1214] Loading model from path /tmp/tmptoc00a2p/model/ with prefix 48117e9477af4844
[INFO 2025-02-05T09:43:11.386870211+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2025-02-05T09:43:11.386916923+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2025-02-05T09:43:11.386943886+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2025-02-05T09:43:12.320670225+00:00 kernel.cc:1214] Loading model from path /tmp/tmps5phdm3f/model/ with prefix 9f877e40a6cf47f6
[INFO 2025-02-05T09:43:12.325636659+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2025-02-05T09:43:13.749801497+00:00 kernel.cc:1214] Loading model from path /tmp/tmpprs_0wrl/model/ with prefix 43cff670f6e14cff
[INFO 2025-02-05T09:43:13.776552916+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2025-02-05T09:43:14.73103253+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmt563brv/model/ with prefix fcba4baa07f041f2
[INFO 2025-02-05T09:43:14.736100206+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2025-02-05T09:43:16.313411896+00:00 kernel.cc:1214] Loading model from path /tmp/tmphf7675lr/model/ with prefix 7c24049700f2434c
[INFO 2025-02-05T09:43:16.345414794+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2025-02-05T09:43:17.308441924+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx29nx2k3/model/ with prefix 8ebf0994a5ac4401
[INFO 2025-02-05T09:43:17.314863474+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2025-02-05T09:43:18.215882627+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvk6y22nq/model/ with prefix 0a691b829b7f402d
[INFO 2025-02-05T09:43:18.21934313+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2025-02-05T09:43:19.270057887+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1x86twza/model/ with prefix ae2e73634ff54f5b
[INFO 2025-02-05T09:43:19.278521226+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2025-02-05T09:43:20.675077961+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcy3r027i/model/ with prefix 1756fc8fb6084986
[INFO 2025-02-05T09:43:20.697353223+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2025-02-05T09:43:21.799078374+00:00 kernel.cc:1214] Loading model from path /tmp/tmprve_ap2v/model/ with prefix 0f3b872f77014e3c
[INFO 2025-02-05T09:43:21.810147341+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:43:21.810192555+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2025-02-05T09:43:23.489706199+00:00 kernel.cc:1214] Loading model from path /tmp/tmp73g5dz65/model/ with prefix ccbdb7449ce14f60
[INFO 2025-02-05T09:43:23.507184573+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2025-02-05T09:43:24.504180944+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvk7yqpha/model/ with prefix 6c9c9ea684c44aab
[INFO 2025-02-05T09:43:24.510720323+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2025-02-05T09:43:25.745379206+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvfyawl15/model/ with prefix f4896c0437c346a2
[INFO 2025-02-05T09:43:25.761949874+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2025-02-05T09:43:26.73271572+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwd47vxzr/model/ with prefix ad04f7b76f2b4b6d
[INFO 2025-02-05T09:43:26.739421587+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2025-02-05T09:43:27.850848397+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxcq8nl84/model/ with prefix 6b6112ca1c8843ad
[INFO 2025-02-05T09:43:27.863140056+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2025-02-05T09:43:28.837090813+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxayt0kmt/model/ with prefix 2612151095e3470e
[INFO 2025-02-05T09:43:28.843938659+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2025-02-05T09:43:29.866557045+00:00 kernel.cc:1214] Loading model from path /tmp/tmpejj7putx/model/ with prefix b1413930487246fb
[INFO 2025-02-05T09:43:29.874902462+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2025-02-05T09:43:31.085613506+00:00 kernel.cc:1214] Loading model from path /tmp/tmpphfu_g3i/model/ with prefix 0311f412f0b640df
[INFO 2025-02-05T09:43:31.100186078+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2025-02-05T09:43:32.324725412+00:00 kernel.cc:1214] Loading model from path /tmp/tmpatp1nqxp/model/ with prefix c0f97cf0d3184704
[INFO 2025-02-05T09:43:32.339329252+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:43:32.339391905+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2025-02-05T09:43:33.486829813+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdikkhit_/model/ with prefix 7c288e6f75f64577
[INFO 2025-02-05T09:43:33.499897866+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2025-02-05T09:43:34.867837786+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfbxhqmjg/model/ with prefix 07073e9be22b435c
[INFO 2025-02-05T09:43:34.88781365+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2025-02-05T09:43:36.171096165+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqp76iehx/model/ with prefix d5483476a9ca464c
[INFO 2025-02-05T09:43:36.187935519+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2025-02-05T09:43:37.197450301+00:00 kernel.cc:1214] Loading model from path /tmp/tmpke6uhrko/model/ with prefix dc8e14d2f696472f
[INFO 2025-02-05T09:43:37.203247942+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2025-02-05T09:43:38.18297694+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8d29iquj/model/ with prefix 127a2c4fa3834830
[INFO 2025-02-05T09:43:38.189421262+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2025-02-05T09:43:39.268381352+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdqc4ywb4/model/ with prefix d14dc1162eca46b3
[INFO 2025-02-05T09:43:39.278145611+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2025-02-05T09:43:40.262322457+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3jtudu7m/model/ with prefix 7c406c29a4e24235
[INFO 2025-02-05T09:43:40.2691337+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2025-02-05T09:43:41.502861752+00:00 kernel.cc:1214] Loading model from path /tmp/tmphtshlnmc/model/ with prefix 2ba7a366a6e1416a
[INFO 2025-02-05T09:43:41.515929516+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2025-02-05T09:43:42.609567117+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2cwsubf7/model/ with prefix 27fc245ca78e4d72
[INFO 2025-02-05T09:43:42.620600108+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:43:42.620659006+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2025-02-05T09:43:43.634395374+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5dr00wg0/model/ with prefix 219ef2b07505456b
[INFO 2025-02-05T09:43:43.642013812+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2025-02-05T09:43:44.625873838+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkm26bi91/model/ with prefix 2ee4de5192d84d1b
[INFO 2025-02-05T09:43:44.631619116+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2025-02-05T09:43:45.909016185+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmk5oco82/model/ with prefix d8a60e18ed4b43f2
[INFO 2025-02-05T09:43:45.929453083+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2025-02-05T09:43:47.568233125+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiscqwl9c/model/ with prefix 51cb70fc25ba46eb
[INFO 2025-02-05T09:43:47.598314534+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2025-02-05T09:43:49.235485731+00:00 kernel.cc:1214] Loading model from path /tmp/tmplolpikkl/model/ with prefix d6911843942340e4
[INFO 2025-02-05T09:43:49.248105513+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2025-02-05T09:43:50.299033129+00:00 kernel.cc:1214] Loading model from path /tmp/tmph_zvt2ln/model/ with prefix d9bc9653e6d840fe
[INFO 2025-02-05T09:43:50.305903179+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2025-02-05T09:43:51.57385459+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkctwf02r/model/ with prefix 64c5ba13049c430f
[INFO 2025-02-05T09:43:51.588717384+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2025-02-05T09:43:52.654149002+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnusagmuh/model/ with prefix ab82fa0c94e24923
[INFO 2025-02-05T09:43:52.66412393+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:43:52.664195972+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2025-02-05T09:43:53.724798696+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcx7opv02/model/ with prefix 3824726ef4d548fa
[INFO 2025-02-05T09:43:53.733895542+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2025-02-05T09:43:54.982961163+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsxe92biz/model/ with prefix 37b528505b6e4f73
[INFO 2025-02-05T09:43:54.999072635+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2025-02-05T09:43:56.099792497+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu8u9cy43/model/ with prefix 8b0899ee056f4e1a
[INFO 2025-02-05T09:43:56.109956192+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2025-02-05T09:43:57.319358125+00:00 kernel.cc:1214] Loading model from path /tmp/tmpghehemk3/model/ with prefix 19845621320e4d21
[INFO 2025-02-05T09:43:57.333979164+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2025-02-05T09:43:58.535475443+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgtdrc2iq/model/ with prefix 466491d88dff4910
[INFO 2025-02-05T09:43:58.54955756+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2025-02-05T09:43:59.484241751+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjezsowio/model/ with prefix a0df57d63dc041cc
[INFO 2025-02-05T09:43:59.489327073+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2025-02-05T09:44:00.7420916+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl0xwd67s/model/ with prefix cb86fdce026f4521
[INFO 2025-02-05T09:44:00.759351709+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2025-02-05T09:44:01.905066068+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3bx9fsyn/model/ with prefix ac4a473aa1ca465f
[INFO 2025-02-05T09:44:01.91534336+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2025-02-05T09:44:03.258558292+00:00 kernel.cc:1214] Loading model from path /tmp/tmpge54vk5b/model/ with prefix c6b160957e614d31
[INFO 2025-02-05T09:44:03.277479052+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:03.277543476+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2025-02-05T09:44:04.352652442+00:00 kernel.cc:1214] Loading model from path /tmp/tmp395zg7e3/model/ with prefix 55fc4bbaa3ab4467
[INFO 2025-02-05T09:44:04.363241003+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2025-02-05T09:44:05.291641042+00:00 kernel.cc:1214] Loading model from path /tmp/tmp55h8lg0z/model/ with prefix 65cf108bc7634ad0
[INFO 2025-02-05T09:44:05.295759634+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2025-02-05T09:44:06.53892525+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc3qdrcue/model/ with prefix 5f50f353804c465b
[INFO 2025-02-05T09:44:06.555331572+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2025-02-05T09:44:07.741540776+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq6sge1v0/model/ with prefix 01756dcaab8f4f89
[INFO 2025-02-05T09:44:07.756256272+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2025-02-05T09:44:08.709642702+00:00 kernel.cc:1214] Loading model from path /tmp/tmprg_rc6z0/model/ with prefix 74f1034fc3d9472c
[INFO 2025-02-05T09:44:08.714797477+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2025-02-05T09:44:09.726350521+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdxjbh8gd/model/ with prefix db0f40a05fbe451c
[INFO 2025-02-05T09:44:09.73375708+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2025-02-05T09:44:10.87956299+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_9cu6dhi/model/ with prefix 81642d1950714e8a
[INFO 2025-02-05T09:44:10.893324698+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2025-02-05T09:44:12.163855971+00:00 kernel.cc:1214] Loading model from path /tmp/tmpasa7xozq/model/ with prefix 1dac5abb86184c4a
[INFO 2025-02-05T09:44:12.181500908+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2025-02-05T09:44:13.214160261+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5la2vgkk/model/ with prefix a24127d993184a35
[INFO 2025-02-05T09:44:13.222760811+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2025-02-05T09:44:14.249666531+00:00 kernel.cc:1214] Loading model from path /tmp/tmpocvlf0o0/model/ with prefix 744ad6dfa8744f28
[INFO 2025-02-05T09:44:14.257960585+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:14.258001533+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2025-02-05T09:44:15.186898054+00:00 kernel.cc:1214] Loading model from path /tmp/tmptig24951/model/ with prefix afba8aecd3b84be7
[INFO 2025-02-05T09:44:15.190808894+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2025-02-05T09:44:16.462939136+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsiy4od8q/model/ with prefix 8d9774ea14824115
[INFO 2025-02-05T09:44:16.479661839+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2025-02-05T09:44:18.225247015+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbnmnh2oq/model/ with prefix 6b73ae29c1d1472b
[INFO 2025-02-05T09:44:18.239637486+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2025-02-05T09:44:19.302988609+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbg4c5c74/model/ with prefix 7e4c510f862e4af1
[INFO 2025-02-05T09:44:19.308290397+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2025-02-05T09:44:20.33557757+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu517ptkt/model/ with prefix 650d4823616d4362
[INFO 2025-02-05T09:44:20.343165102+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2025-02-05T09:44:21.465777857+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmiw_6f9m/model/ with prefix bd01575718dc4c49
[INFO 2025-02-05T09:44:21.476416395+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2025-02-05T09:44:22.584881662+00:00 kernel.cc:1214] Loading model from path /tmp/tmp42btgbed/model/ with prefix f883beda531e4ef3
[INFO 2025-02-05T09:44:22.596056071+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2025-02-05T09:44:23.582681161+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyo2bfpj_/model/ with prefix 427c5687bdf94200
[INFO 2025-02-05T09:44:23.588328214+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2025-02-05T09:44:25.1745254+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjw9boben/model/ with prefix a8494132009b481c
[INFO 2025-02-05T09:44:25.20295924+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:25.20300879+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2025-02-05T09:44:26.290794896+00:00 kernel.cc:1214] Loading model from path /tmp/tmp301l6lnk/model/ with prefix 7b8be401655643a0
[INFO 2025-02-05T09:44:26.300934147+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2025-02-05T09:44:27.347726226+00:00 kernel.cc:1214] Loading model from path /tmp/tmputj6p8ys/model/ with prefix a96677e393f947da
[INFO 2025-02-05T09:44:27.356355965+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2025-02-05T09:44:28.327345917+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkuswv5m_/model/ with prefix a4fb227dddbf4bd5
[INFO 2025-02-05T09:44:28.332744161+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2025-02-05T09:44:29.341999017+00:00 kernel.cc:1214] Loading model from path /tmp/tmplv4lrk5r/model/ with prefix 724d6b2530354911
[INFO 2025-02-05T09:44:29.348966211+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2025-02-05T09:44:30.72473581+00:00 kernel.cc:1214] Loading model from path /tmp/tmphhsyedbl/model/ with prefix 6d80ffa9bcf94f0c
[INFO 2025-02-05T09:44:30.746113919+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2025-02-05T09:44:31.960542085+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyruiqli8/model/ with prefix 9fdb70cc499243c8
[INFO 2025-02-05T09:44:31.977550272+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2025-02-05T09:44:32.999664937+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsd29s64v/model/ with prefix fcb320a6cbc145c6
[INFO 2025-02-05T09:44:33.006970991+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2025-02-05T09:44:34.101713905+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr1n8xm6x/model/ with prefix a3d583cc6f354c42
[INFO 2025-02-05T09:44:34.111974106+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2025-02-05T09:44:35.184211978+00:00 kernel.cc:1214] Loading model from path /tmp/tmps6ay7ttt/model/ with prefix 5e82da7609854ed4
[INFO 2025-02-05T09:44:35.192969839+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2025-02-05T09:44:36.490378589+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqoyqp3jy/model/ with prefix 06ff99f457544f05
[INFO 2025-02-05T09:44:36.507891185+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:36.507940222+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2025-02-05T09:44:37.534960656+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2eavg5hu/model/ with prefix f98ae9b027bc47e9
[INFO 2025-02-05T09:44:37.542847552+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2025-02-05T09:44:38.702933616+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_a05ejqg/model/ with prefix 462521175b12449f
[INFO 2025-02-05T09:44:38.715622517+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2025-02-05T09:44:39.778970461+00:00 kernel.cc:1214] Loading model from path /tmp/tmpezgqu_xl/model/ with prefix 9c5546f6906a4642
[INFO 2025-02-05T09:44:39.787997818+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2025-02-05T09:44:40.780654103+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5p8lt1kq/model/ with prefix e5804ac5a3074b1d
[INFO 2025-02-05T09:44:40.785201913+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2025-02-05T09:44:41.787715802+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8lw96kvz/model/ with prefix da3ab680a0a84cbb
[INFO 2025-02-05T09:44:41.793091578+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2025-02-05T09:44:42.79917182+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv2rxm4ku/model/ with prefix 0801d187f1bb4478
[INFO 2025-02-05T09:44:42.806516744+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2025-02-05T09:44:43.82860199+00:00 kernel.cc:1214] Loading model from path /tmp/tmps5i6q1qr/model/ with prefix 6430d36d107541c2
[INFO 2025-02-05T09:44:43.836621375+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2025-02-05T09:44:44.957840601+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqjitslc6/model/ with prefix c6dd9d67f20a41f0
[INFO 2025-02-05T09:44:44.968712144+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2025-02-05T09:44:46.154836846+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv5ycflqy/model/ with prefix bd618d258b744b7d
[INFO 2025-02-05T09:44:46.16728384+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2025-02-05T09:44:47.310244273+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbtuo6mhh/model/ with prefix a24f5e4de0a14477
[INFO 2025-02-05T09:44:47.320981199+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:47.321028916+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2025-02-05T09:44:48.399377001+00:00 kernel.cc:1214] Loading model from path /tmp/tmp26b_n5ml/model/ with prefix 01c584c6597b4cb9
[INFO 2025-02-05T09:44:48.409172595+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2025-02-05T09:44:50.49702751+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcqhdxj14/model/ with prefix 8f360e00491d40a0
[INFO 2025-02-05T09:44:50.518639899+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2025-02-05T09:44:51.635084273+00:00 kernel.cc:1214] Loading model from path /tmp/tmpslaqepnu/model/ with prefix f1a7d9fc73434eb1
[INFO 2025-02-05T09:44:51.642879055+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2025-02-05T09:44:52.943153198+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwn4zfddz/model/ with prefix d1566a1f805846a3
[INFO 2025-02-05T09:44:52.961592232+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2025-02-05T09:44:54.326773182+00:00 kernel.cc:1214] Loading model from path /tmp/tmph7kymke7/model/ with prefix faebf37e04344e7f
[INFO 2025-02-05T09:44:54.347353024+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2025-02-05T09:44:55.532238344+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4i5grh63/model/ with prefix 97014f973654409e
[INFO 2025-02-05T09:44:55.545407497+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2025-02-05T09:44:56.532716778+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdw0x4jh3/model/ with prefix 574c50fd03fb4e92
[INFO 2025-02-05T09:44:56.538337784+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2025-02-05T09:44:57.672864615+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxd83kg3t/model/ with prefix b642ce5a307a4610
[INFO 2025-02-05T09:44:57.684368633+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:44:57.684416701+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2025-02-05T09:44:58.715885173+00:00 kernel.cc:1214] Loading model from path /tmp/tmpot2zyb3w/model/ with prefix 755f7fd7bd6e4356
[INFO 2025-02-05T09:44:58.723491168+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2025-02-05T09:45:00.057799198+00:00 kernel.cc:1214] Loading model from path /tmp/tmpev27rggy/model/ with prefix acaeac6367df45fc
[INFO 2025-02-05T09:45:00.077685706+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2025-02-05T09:45:01.325829653+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw3jgu61j/model/ with prefix e07bdc66f06544f8
[INFO 2025-02-05T09:45:01.340334509+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2025-02-05T09:45:02.374474045+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgbxlm_wy/model/ with prefix 9db4e71fa14f43ce
[INFO 2025-02-05T09:45:02.381702007+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2025-02-05T09:45:03.466609847+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_rqs0yvi/model/ with prefix 41a38d838e16483f
[INFO 2025-02-05T09:45:03.476087581+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2025-02-05T09:45:04.598795099+00:00 kernel.cc:1214] Loading model from path /tmp/tmpa45vehuz/model/ with prefix 0ed606dcf6b34cf3
[INFO 2025-02-05T09:45:04.610397616+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2025-02-05T09:45:05.599245944+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd4lm5tvy/model/ with prefix b36a1a6b4de94a6c
[INFO 2025-02-05T09:45:05.605029979+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2025-02-05T09:45:06.660200006+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt991r42m/model/ with prefix 825fc31f88a0465a
[INFO 2025-02-05T09:45:06.668856113+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2025-02-05T09:45:07.911088246+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkvflc03y/model/ with prefix e6dfc3622649478e
[INFO 2025-02-05T09:45:07.926953262+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-02-05T09:45:07.927001841+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 